ath79: add support for Netgear PGZNG1 (ADT Pulse Gateway)

riptidewave93 opened the pull request:
The following PR adds 3x commits that are required to add support for the Netgear PGZNG1 (ADT Pulse Gateway).

This is a really unique ar9344 board due to the large amount of RAM and Storage, since the stock OS runs a java applet. It also has onboard Z-Wave and RS422, but neither are working in OpenWRT.

riptidewave93 committed:
kernel: add upstream patches for pca955x led driver

These patches are to support the pca955x led with OpenWRT correctly via
device tree on linux 5.10. Without these, the new LED function/color/reg
features can not be used.

Signed-off-by: Chris Blake <chrisrblake93@gmail.com>

riptidewave93 committed:
kernel: add kmod-leds-pca955x

This patch adds support for the mainline kernel module for the PCA955x
LED driver. Note this requires i2c and GPIO support. Also worth calling
out this driver also enables GPIO support, depending on device tree
configuration.

Signed-off-by: Chris Blake <chrisrblake93@gmail.com>

chunkeey commented on target/linux/ath79/dts/ar9344_netgear_pgzng1.dts:
```diff
@@ -0,0 +1,341 @@
+				function = LED_FUNCTION_BOOT;
+				color = <LED_COLOR_ID_GREEN>;
+				reg = <0>;
+				default-state = "off";
```
The default state is "off". Is there something that switches it on?

chunkeey commented on target/linux/ath79/dts/ar9344_netgear_pgzng1.dts:
```diff
@@ -0,0 +1,341 @@
+		compatible = "gpio-leds";
+
+		eth1_link {
+			label = "green:eth1";
```
please add color + function properties just like the leds above :)

you can use the function-enumerator property to differentiate between either ethernet port.

riptidewave93 replied:
Ill try with LED_FUNCTION_WAN/LED_FUNCTION_LAN first since it's just 2 ports that are labeled.

chunkeey commented on target/linux/ath79/dts/ar9344_netgear_pgzng1.dts:
```diff
@@ -0,0 +1,341 @@
+	status = "okay";
+};
+
+&caldata {
```
please integrate this into the caldata node from above. (you can drop the compatible and #address-cells + #size-cells since they are specified twice)

chunkeey commented on target/linux/ath79/image/nand.mk:
```diff
@@ -231,6 +231,23 @@ define Device/netgear_ath79_nand
+  PAGESIZE := 2048
+  BLOCKSIZE := 128k
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma | \
+    check-size $$(KERNEL_SIZE)
```
The check-size is not needed. For KERNEL, the ```Device/Build/kernel``` function will perform a check-size automatically.

chunkeey commented on target/linux/ath79/image/nand.mk:
```diff
@@ -231,6 +231,23 @@ define Device/netgear_ath79_nand
+  DEVICE_MODEL := PGZNG1
+  DEVICE_VENDOR := NETGEAR
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ledtrig-usbport kmod-i2c-gpio \
+  kmod-leds-pca955x kmod-rtc-isl1208 kmod-spi-dev uboot-envtools
```
one extra indent please (or is it there and this github view hides it?). 

uboot-envtools is part of the DEFAULT_PACKAGES of the ath79 target.

chunkeey commented on target/linux/ath79/image/nand.mk:
```diff
@@ -231,6 +231,23 @@ define Device/netgear_ath79_nand
+define Device/netgear_pgzng1
+  SOC := ar9344
+  DEVICE_MODEL := PGZNG1
+  DEVICE_VENDOR := NETGEAR
```
You could specify an alternate vendor + product through the DEVICE_ALT0_VENDOR + DEVICE_ALT0_MODEL name for the "ADT Pulse Gateway" branding.

chunkeey commented on target/linux/ath79/image/nand.mk:
```diff
@@ -231,6 +231,23 @@ define Device/netgear_ath79_nand
+  BLOCKSIZE := 128k
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma | \
+    check-size $$(KERNEL_SIZE)
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | uImage lzma
```
By dropping the check-size from the KERNEL about, you should be able to drop the (now identical) KERNEL_INITRAMFS line?

chunkeey commented on target/linux/ath79/dts/ar9344_netgear_pgzng1.dts:
```diff
@@ -0,0 +1,341 @@
+			read-only;
+		};
+
+		partition@ff80000 {
```
Hmm

some (really basic) NAND chips expose their bad block table at the end. Could this be padding the bbt (instead of a "dummy partition"?) You can check by looking at the content. If it's all FF then it's likely not used.

riptidewave93 replied:
It's all FF's, but I don't believe it's a bad block table. I looked at the Datasheet for the NAND (SK Hynix h27u2g8f2ctr) and don't see a callout mentioning this.

I am going to leave this in for now, but I believe it is indeed leftover from whatever SDK was used by Netgear in case older NAND packages were used.

chunkeey commented:
Could you please add a "hardware specification" list to the commit which adds support for the device?
(+ a mac-addresses-table like in 41be1a2de2f84f532a7eb1c11ece99811ca47ba5 is also nice)

riptidewave93 commented:
@chunkeey PR is updated to address all of your feedback.

riptidewave93 committed:
ath79: add support for Netgear PGZNG1

This adds support for the Netgear PGZNG1, also known as the ADT Pulse
Gateway.

Hardware:
CPU: Atheros AR9344
Memory: 256MB
Storage: 256MB NAND
USB: 1x USB 2.0
Ethernet: 2x 100Mb/s
WiFi: Atheros AR9340 2.4GHz

UART:
Header marked JPE1. Pinout is VCC, TX, RX, GND. The marked pin, closest
to the JPE1 marking, is VCC. Note VCC isn't required to be connected
for UART to work.

Enable Stock Firmware Shell Access:
1. Interupt u-boot and run the following commands
setenv console_mode 1
saveenv
reset

This will enable a UART shell in the firmware. You can then login using
the root password of `icontrol`. If that doesn't work, the device is
running a firmware based on OpenWRT where you can drop into failsafe to
mount the FS and then modify /etc/passwd.

Installation Instructions:
1. Interupt u-boot and run the following commands
setenv active_image 0
setenv stock_bootcmd nboot 0x81000000 0 \${kernel_offset}
setenv openwrt_bootcmd nboot 0x82000000 0 \${kernel_offset}
setenv bootcmd run openwrt_bootcmd
saveenv

2. boot initramfs image via TFTP u-boot
tftpboot 0x82000000 openwrt-ath79-nand-netgear_pgzng1-initramfs-kernel.bin; bootm 0x82000000

3. Once booted, use LuCI sysupgrade to
flash openwrt-ath79-nand-netgear_pgzng1-squashfs-sysupgrade.bin

MAC Table:
WAN (eth0): xx:xa - caldata 0x0
LAN (eth1): xx:xb - caldata 0x6
WLAN (phy0): xx:xc - burned into ath9k caldata

Not Working:
Z-Wave
RS422

Signed-off-by: Chris Blake <chrisrblake93@gmail.com>

chunkeey commented:
Merged. Thanks!
